cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -657,6 +657,13 @@ else()
            "${THIRDPARTY_MIRROR_URL}/snappy-${ARROW_SNAPPY_BUILD_VERSION}.tar.gz")
 endif()
 
+# Remove these two lines once https://github.com/substrait-io/substrait/pull/342 merges
+set(ENV{ARROW_SUBSTRAIT_URL}
+    "https://github.com/substrait-io/substrait/archive/e59008b6b202f8af06c2266991161b1e45cb056a.tar.gz"
+)
+set(ARROW_SUBSTRAIT_BUILD_SHA256_CHECKSUM
+    "f64629cb377fcc62c9d3e8fe69fa6a4cf326f34d756e03db84843c5cce8d04cd")
```

Comment:
Hmm...this isn't great to be relying on a Substrait fork.  That being said, we aren't going to be compatible with anything until https://github.com/substrait-io/substrait/pull/342 merges.  So I suppose this is just a natural consequence of working with emerging specifications.  It puts us in a better state for the short term.